cEP-0024.md

```diff
@@ -0,0 +1,255 @@
+| Type     | Process                                               |
+
+## Abstract
+
+This cEP describes the improvements to the coala workflow and the necessary
+automation in [GitMate](https://gitmate.io/) to realize it, as a part of the
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5344316227059712).
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there are a number of missing features which would save the time of
+coala community.
+
+Many of the plugins in GitMate are general plugins, but coala needs more
+plugins which support coala development workflow. So, this project
+adds plugins to auto-review, auto-reject, auto-label PRs, assign, unassign
+issues, squash commits.
+
+## Changes in coala workflow
+
+### Use increasing difficulty level order to assign issues
+
+The current assign process in coala is to check team membership of a user,
+difficulty level of the issue and assign the issue to user if eligible. To
+implement the current assign process in GitMate, IGitt would need to support
+teams.
+
+Currently, IGitt doesn't support teams and there are no teams in GitLab.
+Considering [GitLab groups](https://docs.gitlab.com/ee/user/group/) and
+[GitHub teams](https://help.github.com/articles/about-teams/) as teams in
+IGitt is a possible solution for IGitt implementation. But, it requires extra
+tools to sync GitHub team and GitLab group memberships, which is a manual
+process and isn't automated yet.
+
+Also, team membership information is not publicly accessible in GitHub, so any
+system which needs to implement the current assign system would require
+organization authentication tokens.
+
+The best possible solution would be using an increasing difficulty level order,
+so that a user can be assigned to an issue only if the user has solved issues
+of lower difficulty level before requesting to get assigned to a higher
+difficulty level issue.
+
+#### Drawbacks
+
+If a user completed the lower difficulty level issues in GitHub and want to
+work on a higher difficulty level issue in GitLab, he/she has to repeat the
+newcomer process in GitLab. To resolve this issue, GitMate needs to find a way
+to link GitHub org and GitLab org, this would be easy when they have the same
+name, as in the case of `coala`. But, resolving this issue is out of scope for
+this project.
+
+If a user is worthy enough to work on the higher level difficulty issues
+directly, then that user needs to be manually assigned by the maintainers until
+that user completes at least one low difficulty issue. Even with current assign
+system, manual intervention by maintainers is needed to add a user to
+developers team.
+
+## New Features
+
+### Auto-assign issue when it is mentioned in any of PR commits
+
+When a PR is opened or updated, if any of PR's commit message referenced
+any issues with keywords `Fixes` or `Closes`, there is no
```

Comment:
For context, the "there is no CI failure in PR" condition here was because I mentioned in an earlier review that we probably dont want an issue auto-assigned to a person if they create a crappy commit which fails CI.

To prevent this,  would mean that some part of the plugin needs to run after CI checks have had status' returned as successful or failure.

When the plugin runs needs to be stated explicitly and clearly stated here.

Currently "When a PR is opened or updated" is not actually possible as "there is no CI failure in PR" requires a "When CI on a PR has passed"..

Maintainers will be wondering why it hasnt happened automatically when the PR is created.

A problem which this creates is: which CI must pass or not fail?

This plugin can not wait for all CI to pass, because some GitMate plugins do not return a status until the commits have been `ack`'d, which could be many days after the PR has been created.  Should the plugin wait for all CI checks except the ack plugin?

However note that my previous comment about crappy commits was when I believed this plugin would work on commits *before* they create a PR (which is what was in the project description).

If this plugin will only work on PRs , perhaps it should not wait for CI to pass.  Otherwise in a "workshop" situation with lots of newcomers trying to find issues to work on, lots of PRs will be created for the same issues because none of the PRs pass CI , and the order of https://coala.io/newcomer is not randomised , so the attendees typically go for the issues on the first few pages.

If they have already created a PR, and it is a newcomer issue, it is probably better to just assign it to them even if it fails CI, in order to avoid other people being assigned to that issue when we have a PR which fails CI.   We dont want to incentivise "First PR to pass CI gets to claim the issue".  That creates a big mess for maintainers to cleanup. 

Again assuming this plugin will only work on PRs, another alternative is to auto close the PR if it is a newcomer issue and they were not assigned first.  That forces an orderly approach to newcomer issues.  Note this excludes other types of difficulty because we do not want to auto-close PRs from random developers who are very experienced.

Reply:
>This plugin can not wait for all CI to pass, because some GitMate plugins do not return a status until the commits have been ack'd, which could be many days after the PR has been created. Should the plugin wait for all CI checks except the ack plugin?

This is the reason I said "there is no CI failure in PR". If CI is in pending state, it will not be considered as failure, else if there is a CI failure, it will not be assigned.

>If this plugin will only work on PRs , perhaps it should not wait for CI to pass. Otherwise in a "workshop" situation with lots of newcomers trying to find issues to work on, lots of PRs will be created for the same issues because none of the PRs pass CI , and the order of https://coala.io/newcomer is not randomised , so the attendees typically go for the issues on the first few pages.

Yeah, maybe workshop situation is different. But it isn't of much importance right now, this mechanism can be implemented later.